### PR TITLE
Now returning an error package when receiving the wrong block number

### LINKED ISSUE #35: Replace the assert statement in connection::Connection::put with an error packet and terminate the transmission
A misbehaving peer could send us an ack with the wrong block number; however, panicking via an assert is not a helpful way to deal with this.

We don't need to go out of our way to try to get the connection back on track. Create a courtesy Error packet, send it to the peer, and then return early from this method with an err.

### PR DESCRIPTION
**Pull request by 1c3t3a**
Closes #35. The connection::put Method now sends an error package before ending the communication in case the current_block count is not plausible.

### CODE REVIEW

**Comment by 1c3t3a** [comment]
I dont't really get how he's not finding [std::io::Error](https://doc.rust-lang.org/std/io/struct.Error.html) in the [error log](https://travis-ci.com/github/connorkuehl/tftp/jobs/395267498).

**Comment by connorkuehl** [comment]
@1c3t3a it's because your development branch is out-of-date with my branch and that import statement might have disappeared since then. No worries! This is easily fixed by rebasing on top of my branch.

When you're on your local branch:

```console
$ git fetch https://github.com/connorkuehl/tftp main
$ git rebase FETCH_HEAD
```

There might be some merge conflicts that need to be resolved.

**Comment by 1c3t3a** [comment]
> @1c3t3a it's because your development branch is out-of-date with my branch and that import statement might have disappeared since then. No worries! This is easily fixed by rebasing on top of my branch.
> 
> When you're on your local branch:
> 
> ```
> $ git fetch https://github.com/connorkuehl/tftp main
> $ git rebase FETCH_HEAD
> ```
> 
> There might be some merge conflicts that need to be resolved.

Ah I see! I rebased and added the import again, now it should work fine!

File: src/connection.rs [hunk 1]
```diff
@@ -69,7 +69,14 @@ impl Connection {
 
             let ack: Packet<Ack> = self.socket.expect_packet(&buf[..bytes_recvd])?;
 
-            assert_eq!(Block::new(current_block), ack.body.block);
+            if Block::new(current_block) != ack.body.block {
+                let error = Packet::<Error>::error(
+                    Code::IllegalOperation,
+                    "Got an ack with the wrong packet number, terminate excecution",
+                );
```
Comment (connorkuehl) [inline src/connection.rs:76]:
```suggestion
                let error = Packet::error(
                    Code::IllegalOperation,
                    format!("expected ACK for {} but got ACK for {}", current_block, ack.body.block),
                );
```

Awesome! Just a couple of tweaks. I don't think we have to turbofish `Packet::<Error>::error`, I think Rust knows which `Packet<T>` we mean when we use `Packet::error`.

I tweaked the error message to include more info.

I haven't tried compiling these suggestions so maybe a bit of massaging is necessary.

**Review by connorkuehl — CHANGES_REQUESTED** [review]

**Review by connorkuehl — APPROVED** [review]
Thank you so much! :slightly_smiling_face:

**Pull request merged**